src/pam_krb5_cchelper.8.in
.TH pam_krb5_cchelper 8 2012/07/20 "@OS_DISTRIBUTION@" "System Administrator's Manual"

.SH NAME
@SECURITYDIR@/@PACKAGE@/pam_krb5_cchelper \- Credential cache helper

.SH SYNOPSIS
.B pam_krb5_cchelper [-c|-u|-d] [ccname] [uid] [gid]

.SH DESCRIPTION
The pam_krb5.so module uses pam_krb5_cchelper to create, update, and remove
credential caches.  If it is asked to create a FILE: or DIR:-type credential
cache in a subdirectory of \fI/run/user\fP, it will attempt to create that
subdirectory if it does not already exist.

.SH ARGUMENTS
.IP -c|-u|-d
A flag indicating whether the helper is expected to create, update, or
destroy a ccache.  When creating a ccache, the \fIccname\fP argument
should be a pattern ending in XXXXXX.  When updating or deleting a ccache,
the \fIccname\fP argument should be the name of an extant ccache.

.IP ccname
A credential cache name or name pattern of the form
\fITYPE\fR:\fI@default_ccache_dir@/foo or
\fITYPE\fR:\fI@default_ccache_dir@/fooXXXXXX\fR.  If

.IP uid
An optional numeric UID which the helper will attempt to switch to before
creating a ccache.  The helper continues in its task if the attempt fails.

.IP gid
An optional numeric GID which the helper will attempt to switch to before
creating a ccache.  The helper continues in its task if the attempt fails.

.SH OPERATION
The helper will read contents from its standard input.
.br
If input of suitable length is not read, the specified credential cache is
deleted.
.br
If input of suitable length is read, a temporary file is created and the input
is stored to the file.  If TYPE is FILE, the file's name will be based on the
pattern and the name of this new credential cache will be printed.  If TYPE is
not FILE, a new credential cache will be created and the contents of the
temporary file will be treated as a credential cache whose contents will be
copied into the new credential cache, whose name will be printed.

.SH FILES
\fI@SECURITYDIR@/pam_krb5.so\fR
.br

.SH "SEE ALSO"
.BR mkstemp (3)
.BR pam_krb5 (5)
.BR pam_krb5 (8)
.br

.SH BUGS
Probably, but let's hope not.  If you find any, please file them in the
bug database at http://bugzilla.redhat.com/ against the "pam_krb5" component.

.SH AUTHOR
Nalin Dahyabhai <nalin@redhat.com>


NEWS
- 2.4.1: * handle creation of /run/user/XXX for FILE: and DIR: caches
- 2.4.0: * drop configuration settings that duplicated library settings
	 * drop the existing_ticket option
	 * drop krb4 support
	 * add support for preserving configuration information in ccaches
	 * add support for creating and cleaning up DIR: ccaches
	 * finish cleaning up KEYRING: ccaches
	 * add experimental "armor" and "armor_strategy" options
- 2.3.14:* also drop privileges when reinitializing or refreshing credentials,
           for the sake of login (#822493)
- 2.3.13:* don't bother creating a v5 ccache in "external" mode
	 * add a "trace" option to enable libkrb5 tracing, if available
	 * avoid trying to get password-change creds twice
	 * use an in-memory ccache when obtaining tokens using v5 creds
	 * turn off creds==session in "sshd"
- 2.3.12:* add a "validate_user_user" option to control trying to perform
	   user-to-user authentication to validate TGTs when a keytab is not
	   available
	 * add an "ignore_k5login" option to control whether or not the module
	   will use the krb5_kuserok() function to perform additional
	   authorization checks
	 * turn on validation by default - verify_ap_req_nofail controls how we
	   treat errors reading keytab files now
	 * add an "always_allow_localname" option when we can use
	   krb5_aname_to_localname() to second-guess the krb5_kuserok() check
	 * prefer krb5_change_password() to krb5_set_password()
- 2.3.11:* create credentials before trying to look up the location of the
	   user's home directory via krb5_kuserok()
- 2.3.10:* fine-tune the logic for selecting which key we use for validating
	   credentials
- 2.3.9: * add a "multiple_ccaches" option to allow forcing the previous
	   behavior of not deleting an old ccache whenever we create a new
	   one, but saving them until the call that caused us to create
	   them is reversed
- 2.3.8: * add a "chpw_prompt" option to allow password changes to happen
	   during what the calling application thinks is just a password
	   check, to work around applications that don't handle the case
	   of an expired password correctly (#509092, based on patch from
	   Olivier Fourdan)
- 2.3.7: * when refreshing credentials, store the new creds in the default
	   ccache if $KRB5CCNAME isn't set (#507984)
- 2.3.6: * prefer a "host" key, if one is found, when validating TGTs
	   (#450776)
- 2.3.5: * make prompting behavior for non-existent accounts and users who
	   just press enter match up with those who aren't/don't (#502602,
	   CVE-2009-1384)
- 2.3.4: * don't request password-changing credentials using the same options
	   we use for ticket-granting tickets
- 2.3.3: * close a couple of open pipes to defunct processes, fix a couple
	   of debug messages
- 2.3.2: * fix ccache permissions bypass when the "existing_ticket" option is
	   used (CVE-2008-3825, which affects 2.2.0-2.2.25, 2.3.0, and 2.3.1)
- 2.3.1: * make afs5log's -n option actually work the "null_afs" option
	 * translations for messages!
- 2.3.0: * added the ability to set up tokens in the rxk5 format
	 * added the "token_strategy" option to control which methods we'll
	   try to use for setting tokens
	 * merge "null_afs" functionality from Jan Iven
- 2.2.23: * when we're changing passwords, force at least one attempt to
	    authenticate using the KDC, even in the pathological case where
	    there's no previously- entered password and we were told not to ask
	    for one (#400611)
- 2.2.22: * moved .k5login checks to a subprocess to avoid screwing with the
	    parent process's tokens and PAG (fallout from #371761)
	  * all options which took true/false before ("debug", "tokens", and
	    so on) can now take service names
- 2.2.21: * fix permissions problems on keyring ccaches, so that users can write
	    to them after we've set them up, and we can still do the cleanup
	  * fix permission problems accessing .k5login files in home directories
	    which live in AFS (#371761)
- 2.2.20: * fixes for credential refreshing
	  * avoid running afoul of SELinux policy when attempting to get tokens
- 2.2.19: * the "keytab" option can now be used to specify a custom location
	    for a given service from within krb5.conf
	  * log messages are now logged with facility LOG_AUTHPRIV (or LOG_AUTH
	    if LOG_AUTHPRIV is not defined) instead of the application's default
	    or LOG_USER
	  * added the "pkinit_identity" option to provide a way to specify
	    where the user's public-key credentials are, and "pkinit_flags" to
	    specify arbitrary flags for libkrb5 (Heimdal only)
	  * added the "preauth_options" option to provide a way to specify
	    arbitrary preauthentication options to libkrb5 (MIT only)
	  * added the "ccname_template" option to provide a way to specify
	    where the user's credentials should be stored, so that KEYRING:
	    credential caches can be deployed at will.
- 2.2.18: * fix permissions-related problems creating v4 ticket files
- 2.2.17: * corrected a typo in the pam_krb5(8) man page
	  * clarified that the "tokens" flag should only be needed for
	    applications which are not using PAM correctly
	  * clarified COPYING and .spec file to better reflect licensing as
	    indicated in the source files
- 2.2.16: * don't bother using a helper for creating v4 ticket files when we're
	    just getting tokens
	  * clean up the debug message which we emit when we do v5->v4
	    principal name conversion
	  * compilation fixes
- 2.2.15: * let default "external" and "use_shmem" settings be specified at
	    compile-time
	  * correctly return a "unknown user" error when attempting to change
	    a password for a user who has no corresponding principal (#235020)
	  * don't bother using a helper for creating ccache files, which we're
	    just going to delete, when we need to get tokens
- 2.2.14: * handle "client revoked" errors
- 2.2.13: * make it possible to have more than one ccache (and tktfile) at a

src/pam_krb5_cchelper.c
/*
 * Copyright 2012 Red Hat, Inc.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, and the entire permission notice in its entirety,
 *    including the disclaimer of warranties.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote
 *    products derived from this software without specific prior
 *    written permission.
 *
 * ALTERNATIVELY, this product may be distributed under the terms of the
 * GNU Lesser General Public License, in which case the provisions of the
 * LGPL are required INSTEAD OF the above restrictions.
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN
 * NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
 * USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "../config.h"

#include <sys/types.h>
#include <sys/stat.h>
#include <dirent.h>
#include <errno.h>
#include <fcntl.h>
#include <grp.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include KRB5_H

#ifdef HAVE_KEYUTILS_H
#include <keyutils.h>
#endif

#ifdef USE_SELINUX
#include <selinux/selinux.h>
#include <selinux/label.h>
#endif

#include "xstr.h"

#define PATH_SEPARATOR_S "/"
#define TMP_RUNTIME "/tmp"
#define TMP_PREFIX TMP_RUNTIME PATH_SEPARATOR_S
#define USER_RUNTIME "/run/user"
#define USER_PREFIX USER_RUNTIME PATH_SEPARATOR_S

static int
unlabeled_mkdir(const char *path, mode_t perms, uid_t uid, gid_t gid)
{
	int ret;
	ret = mkdir(path, perms);
	if (ret == 0) {
		if (chown(path, uid, gid) == 0) {
			ret = -1;
		}
	}
	return ret;
}

#ifdef USE_SELINUX
static int
labeled_mkdir(const char *path, mode_t perms, uid_t uid, gid_t gid)
{
	struct selabel_handle *labels;
	security_context_t context;
	int ret;

	if (!is_selinux_enabled()) {
		return unlabeled_mkdir(path, perms, uid, gid);
	}

	ret = -1;
	labels = selabel_open(SELABEL_CTX_FILE, NULL, 0);
	if (labels != NULL) {
		memset(&context, 0, sizeof(context));
		if (selabel_lookup(labels, &context, path, S_IFDIR) == 0) {
			if (setfscreatecon(context) == 0) {
				ret = unlabeled_mkdir(path, perms, uid, gid);
				setfscreatecon(NULL);
			}
		} else {
			ret = unlabeled_mkdir(path, perms, uid, gid);
		}
		selabel_close(labels);
	}

	return ret;
}
#else
static int
labeled_mkdir(const char *path, mode_t perms, uid_t uid, gid_t gid)
{
	return unlabeled_mkdir(path, perms, uid, gid);
}
#endif

static int
leading_mkdir(const char *path)
{
	char target[PATH_MAX], *p, *component;
	struct passwd *pwd;
	mode_t saved_umask;
	int ret;
	long uid;

	/* Leading directories may be world-writable. */
	saved_umask = umask(0);
	/* Now, we're only doing this if we know about a given prefix. */
	ret = -1;
#if 0
	if (strncmp(path, TMP_PREFIX, strlen(TMP_PREFIX)) == 0) {
		/* We "know" how to create /tmp. */
		ret = labeled_mkdir(TMP_RUNTIME,
				    S_ISVTX | S_IRWXU | S_IRWXG | S_IRWXO,
				    0, 0);
	} else
#endif
	if (strncmp(path, USER_PREFIX, strlen(USER_PREFIX)) == 0) {
		p = NULL;
		pwd = NULL;
		/* We "know" how to create /run/user/XXX .*/
		snprintf(target, sizeof(target), "%s", path);
		component = target + strlen(USER_PREFIX);
		component[strcspn(component, "/")] = '\0';
		uid = strtol(component, &p, 10);
		if ((uid != LONG_MIN) && (uid != LONG_MAX) &&
		    (p != NULL) && (p != component) && (*p == '\0')) {
			pwd = getpwuid(uid);
		} else {
			if (strlen(component) > 0) {
				pwd = getpwnam(component);
			}
		}
		if (pwd != NULL) {
			ret = labeled_mkdir(target, S_IRWXU,
					    pwd->pw_uid, pwd->pw_gid);
		}
	}
	/* Okay, reset the umask. */
	umask(saved_umask);
	return ret;
}

/* A simple (hopefully) helper which creates a file using mkstemp() and a
 * supplied pattern, attempts to set the ownership of that file, stores
 * whatever it reads from stdin in that file, and then prints the file's name
 * on stdout.
 *
 * While all of this can be done directly by pam_krb5, we need to do it after
 * an exec() to have the file created with the proper context if we're running
 * in an SELinux environment, so the helper is used.  To simplify debugging and
 * maintenance, use of this helper is not conditionalized. */
int
main(int argc, const char **argv)
{
	krb5_context ctx = NULL;
	krb5_ccache ccache = NULL, tmp_ccache = NULL;
	krb5_principal client = NULL;
	char *ccname, *p, input[128 * 1024], pattern[PATH_MAX];
	struct dirent **dents = NULL;
	long long uid, gid;
	gid_t current_gid;
	long id;
	int fd, i, j, n_written, c_flag = 0, d_flag = 0, u_flag = 0;
	size_t n_input, n_output;

	/* Get this out of the way. */
	umask(S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH);

	/* We're not intended to be set*id! */
	if ((getuid() != geteuid()) || (getgid() != getegid())) {
		return 1;
	}

	/* Two or four arguments.  No more, no less, else we bail. */
	if ((argc != 3) && (argc != 5)) {
		return 2;
	}

	/* Check what mode we're in. */
	if (strcmp(argv[1], "-c") == 0) {
		c_flag++;
	} else
	if (strcmp(argv[1], "-d") == 0) {
		d_flag++;
	} else
	if (strcmp(argv[1], "-u") == 0) {
		u_flag++;
	} else {
		return 3;
	}

	/* We'll need a writable string for use as the template. */
	ccname = xstrdup(argv[2]);
	if ((ccname == NULL) || (strchr(ccname, ':') == NULL)) {
		return 4;
	}

	/* Parse the UID, if given. */
	if (argc > 3) {
#ifdef HAVE_STRTOLL
		uid = strtoll(argv[3], &p, 0);
#else
		uid = strtol(argv[3], &p, 0);
#endif
		if ((p == NULL) || (*p != '\0')) {
			return 5;
		}
	} else {
		uid = getuid();
	}

	/* Parse the GID, if given. */
	if (argc > 4) {
#ifdef HAVE_STRTOLL
		gid = strtoll(argv[4], &p, 0);
#else
		gid = strtol(argv[4], &p, 0);
#endif
		if ((p == NULL) || (*p != '\0')) {
			return 6;
		}
	} else {
		gid = getgid();
	}

	/* Attempt to drop supplemental groups and become the given user (if
	 * one was given).  Note that this may all fail if we're unprivileged,
	 * and that is expressly allowed. */
	current_gid = getgid();
	if (getuid() == 0) {
		if (setgroups(0, &current_gid) == -1) {
			if (geteuid() == 0) {
				return 6;
			}
		}
	}
	if (getgid() != gid) {
		if (setregid(gid, gid) == -1) {
			if (geteuid() == 0) {
				return 6;
			}
		}
	}
	if (getuid() != uid) {
		if (setreuid(uid, uid) == -1) {
			if (geteuid() == 0) {
				return 5;
			}
		}
	}

	/* Read stdin. */
	n_input = 0;
	while (n_input < sizeof(input)) {
		i = read(STDIN_FILENO, input + n_input,
			 sizeof(input) - n_input);
		if (i < 0) {
			return 7;
		}
		n_input += i;
		if (i == 0) {
			close(STDIN_FILENO);
			break;
		}
	}
	if (n_input == sizeof(input)) {
		return 8;
	}

	i = krb5_init_context(&ctx);
	if (i != 0) {
		return i;
	}

	/* We have three modes.  First, zero-length input should put us in to
	 * delete mode. */
	if (n_input == 0) {
		if (!d_flag) {
			return 9;
		}
		if (strstr(ccname, "XXXXXX") != NULL) {
			return 9;
		}
		/* The first argument is a ccache to be destroyed. */
		i = krb5_cc_resolve(ctx, ccname, &ccache);
		if (i != 0) {
			krb5_free_context(ctx);
			return i;
		}
		i = krb5_cc_destroy(ctx, ccache);
		/* Some ccache types require a bit more work. */
		if ((i == 0) && (strncmp(ccname, "DIR:", 4) == 0)) {
			if ((j = scandir(ccname + 4, &dents,
					 NULL, &alphasort)) > 0) {
				while (j > 0) {
					if ((strcmp(dents[j - 1]->d_name,
						    ".") != 0) &&
					    (strcmp(dents[j - 1]->d_name,
						    "..") != 0) &&
					    (snprintf(pattern, sizeof(pattern),
						      "%s/%s", ccname + 4,
						      dents[j - 1]->d_name) <
					     (int) sizeof(pattern))) {
						unlink(pattern);
					}
					j--;
				}
			}
			rmdir(ccname + 4);
			/* Nothing we can do if this fails. */
		} 
#ifdef HAVE_KEYUTILS_H
		if ((i == 0) && (strncmp(ccname, "KEYRING:", 8) == 0)) {
			id = keyctl_search(KEY_SPEC_SESSION_KEYRING,
					   "keyring", ccname + 8, 0);
			if (id != (long) -1) {
				id = keyctl_unlink(KEY_SPEC_SESSION_KEYRING,
						   id);
				/* Nothing we can do if this fails. */
			}
		}
#endif
		krb5_free_context(ctx);
		return i;
	}

	/* Non-zero-length input puts us in either create or update mode. */
	if (!c_flag && !u_flag) {
		return 9;
	}

	/* Simplest is if we're being asked to either create or update a FILE
	 * ccache. */
	if (strncmp(ccname, "FILE:", 5) == 0) {
		if (strstr(ccname, "XXXXXX") != NULL) {
			/* Check that we're in create mode, and create
			 * the file. */
			if (!c_flag) {
				return 9;
			}
			fd = mkstemp(ccname + 5);
			if ((fd == -1) && (errno == ENOENT)) {
				fd = errno;
				if (leading_mkdir(ccname + 5) == 0) {
					strcpy(ccname, argv[2]);
					fd = mkstemp(ccname + 5);
				} else {
					errno = fd;
					fd = -1;
				}
			}
		} else {
			/* Check that we're in update mode. */
			if (!u_flag) {
				return 9;
			}
			fd = open(ccname + 5, O_WRONLY | O_TRUNC);
		}
		if (fd == -1) {
			fd = errno;
			krb5_free_context(ctx);
			return fd;
		}
		/* Write the ccache contents to the file. */
		n_output = 0;
		while (n_output < n_input) {
			i = write(fd, input + n_output,
				  n_input - n_output);
			if (i < 0) {
				unlink(ccname + 5);
				krb5_free_context(ctx);
				return 10;
			}
			n_output += i;
		}
		close(fd);
		printf("%s\n", ccname);
		return 0;
	}

	/* Create a temporary file to deserialize the ccache. */
	snprintf(pattern, sizeof(pattern), "FILE:%s/pam_krb5_XXXXXX",
		 getenv("TMPDIR") ?: "/tmp");
	fd = mkstemp(pattern + 5);
	if (fd == -1) {
		krb5_free_context(ctx);
		return 11;
	}
	n_output = 0;
	while (n_output < n_input) {
		i = write(fd, input + n_output, n_input - n_output);
		if (i < 0) {
			krb5_free_context(ctx);
			unlink(pattern + 5);
			close(fd);
			return 12;
		}
		n_output += i;
	}
	close(fd);

	/* Open the file as a ccache. */
	i = krb5_cc_resolve(ctx, pattern, &tmp_ccache);
	if (i != 0) {
		unlink(pattern + 5);
		krb5_free_context(ctx);
		return i;
	}
	i = krb5_cc_get_principal(ctx, tmp_ccache, &client);
	if (i != 0) {
		krb5_cc_destroy(ctx, tmp_ccache);
		krb5_free_context(ctx);
		return i;
	}

	/* If the ccache is a directory, create one, if need be. */
	if (strncmp(ccname, "DIR:", 4) == 0) {
		if ((p = strstr(ccname, "XXXXXX")) != NULL) {
			/* Check that we're in create mode, and create
			 * a directory. */
			if (!c_flag) {
				krb5_cc_destroy(ctx, tmp_ccache);
				krb5_free_context(ctx);
				return 9;
			}
			do {
				/* Try to create a unique directory. */
				strcpy(ccname, argv[2]);
				mktemp(ccname + 4);
				if (strlen(ccname + 4) == 0) {
					i = EINVAL;
				} else {
					i = mkdir(ccname + 4, S_IRWXU);
					if ((i == -1) && (errno == ENOENT)) {
						i = errno;
						if (leading_mkdir(ccname + 4) == 0) {
							i = mkdir(ccname + 4,
								  S_IRWXU);
						} else {
							errno = i;
							i = -1;
						}
					}
				}
			} while ((i != 0) && (errno == EEXIST));
			if (i != 0) {
				krb5_cc_destroy(ctx, tmp_ccache);
				krb5_free_context(ctx);
				return i;
			}
		} else {
			/* Check that we're in update mode. */
			if (!u_flag) {
				krb5_cc_destroy(ctx, tmp_ccache);
				krb5_free_context(ctx);
				return 9;
			}
		}
	} else if (strncmp(ccname, "SCC:", 4) == 0) {
		if ((p = strstr(ccname, "XXXXXX")) != NULL) {
			/* Check that we're in create mode, and create an empty
			 * file which sqlite has no qualms about overwriting. */
			if (!c_flag) {
				krb5_cc_destroy(ctx, tmp_ccache);
				krb5_free_context(ctx);
				return 9;
			}
			fd = mkstemp(ccname + 4);
			if (fd == -1) {
				krb5_cc_destroy(ctx, tmp_ccache);
				krb5_free_context(ctx);
				return i;
			}
			close(fd);
		} else {
			/* Check that we're in update mode. */
			if (!u_flag) {
				krb5_cc_destroy(ctx, tmp_ccache);
				krb5_free_context(ctx);
				return 9;
			}
		}
#ifdef HAVE_KEYUTILS_H
	} else if (strncmp(ccname, "KEYRING:", 8) == 0) {
		if ((p = strstr(ccname, "XXXXXX")) != NULL) {
			/* Check that we're in create mode, and create
			 * a new keyring. */
			if (!c_flag) {
				krb5_cc_destroy(ctx, tmp_ccache);
				krb5_free_context(ctx);
				return 9;
			}
			do {
				/* Try to create a unique keyring name. */
				strcpy(ccname, argv[2]);
				mktemp(ccname + 8);
				if (strlen(ccname + 8) == 0) {
					i = EINVAL;
				} else {
					id = keyctl_search(KEY_SPEC_SESSION_KEYRING,
							   "keyring",
							   ccname + 8, 0);
					if (id == (long) -1) {
						id = add_key("keyring",
							     ccname + 8,
							     NULL, 0,
							     KEY_SPEC_SESSION_KEYRING);
						if (id == (long) -1) {
							break;
						}
					} else {
						errno = EEXIST;
						i = -1;
					}
				}
			} while ((i != 0) && (errno == EEXIST));
			if (i != 0) {
				krb5_cc_destroy(ctx, tmp_ccache);
				krb5_free_context(ctx);
				return i;
			}
		} else {
			/* Check that we're in update mode. */
			if (!u_flag) {
				krb5_cc_destroy(ctx, tmp_ccache);
				krb5_free_context(ctx);
				return 9;
			}
		}
#endif
	} else {
		/* Unsupported ccache type. */
		krb5_cc_destroy(ctx, tmp_ccache);
		krb5_free_context(ctx);
		return 13;
	}

	/* Copy the credentials from the temporary ccache to the
	 * ready-to-receive-them destination. */
	i = krb5_cc_resolve(ctx, ccname, &ccache);
	if (i != 0) {
		krb5_cc_destroy(ctx, tmp_ccache);
		krb5_free_context(ctx);
		return i;
	}
	i = krb5_cc_initialize(ctx, ccache, client);
	krb5_free_principal(ctx, client);
	if (i != 0) {
		krb5_cc_destroy(ctx, ccache);
		krb5_cc_destroy(ctx, tmp_ccache);
		krb5_free_context(ctx);
		return i;
	}
#ifdef HAVE_KRB5_CC_COPY_CREDS
	i = krb5_cc_copy_creds(ctx, tmp_ccache, ccache);
#else
	{
		krb5_creds creds;
		krb5_cc_cursor cursor;
		if ((i = krb5_cc_start_seq_get(ctx, tmp_ccache, &cursor)) != 0) {
			krb5_cc_destroy(ctx, ccache);
			krb5_cc_destroy(ctx, tmp_ccache);
			krb5_free_context(ctx);
			return i;
		}
		memset(&creds, 0, sizeof(creds));
		while ((i = krb5_cc_next_cred(ctx, tmp_ccache, &cursor, &creds)) == 0) {
			krb5_cc_store_cred(ctx, ccache, &creds);
			krb5_free_cred_contents(ctx, &creds);
			memset(&creds, 0, sizeof(creds));
		}
		krb5_cc_end_seq_get(ctx, tmp_ccache, &cursor);
	}
#endif
	if (i != 0) {
		krb5_cc_destroy(ctx, ccache);
		krb5_cc_destroy(ctx, tmp_ccache);
		krb5_free_context(ctx);
		return i;
	}
	krb5_cc_close(ctx, ccache);
	krb5_cc_destroy(ctx, tmp_ccache);
	krb5_free_context(ctx);
	printf("%s\n", ccname);
	return 0;
}
